tee/rpc/node.go

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package rpc
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"net"
+	"net/rpc"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/perun-network/erdstall/tee"
+
+	"perun.network/go-perun/log"
+	"perun.network/go-perun/pkg/sync"
+	"perun.network/go-perun/pkg/sync/atomic"
+)
+
+// Node is a server that exposes an Enclave as a slave.
+type Node struct {
+	enclave tee.Enclave // the slave enclave.
+	running atomic.Bool // whether the server has started.
+	server  *rpc.Server // accepts new connections.
+	stopped sync.Closer // whether the server was commanded to stop.
+}
+
+// NewNode creates a new server which is not yet running.
+func NewNode(impl tee.Enclave) *Node {
+	return &Node{
+		enclave: impl,
+		server:  rpc.NewServer(),
+	}
+}
+
+// Start starts the server and sets its listener. Call only once. Starts listening
+// for new connections in the background.
+func (n *Node) Start(l net.Listener) {
+	if l == nil {
+		log.Panic("Node.Start(): nil listener")
+	}
+	if !n.running.TrySet() || n.IsStopped() {
+		log.Panic("Node.Start(): already running")
+	}
+	n.stopped.OnCloseAlways(func() { l.Close() })
+	if err := n.server.Register(n); err != nil {
+		log.Panicf("Node.Start() registering server: %v", err)
+	}
+	go n.server.Accept(l)
+}
+
+// Void indicates that a function has no arguments o no return value.
+type Void struct{}
```

Comment:
Do we need this exported?

Reply:
Sadly, yes. Go RPC does not allow unexported types in arguments of service functions.